staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go

```diff
@@ -69,6 +69,22 @@ type Encoder interface {
 	Identifier() Identifier
 }
 
+// MemoryAllocator is responsible for allocating memory.
+// It exists to make the cost of object serialization cheaper. In some cases, it allows for allocating memory only once and then reusing it.
+// This approach puts less load on GC and leads to less fragmented memory in general.
+type MemoryAllocator interface {
+	// Allocate reserves memory for n bytes
+	Allocate(n uint64) []byte
+}
+
+// EncoderAllocator  serializes objects in a way that allows callers to manage any additional memory allocations.
+type EncoderAllocator interface {
+	// EncodeWithAllocator writes an object to the provided writer.
```

Comment:
How about explicitly nesting Encoder interface here too?

Reply:
we could but it seems to be redundant atm

Reply:
I think the Identifier() method is the important one for caching - and nesting it makes it more explicit.